torch/lib/THD/base/init_methods/InitMethodFile.cpp

```diff
@@ -88,47 +89,105 @@ InitMethod::Config initFile(std::string file_path, rank_type world_size, std::st
 
-  config.rank = std::count(content.begin(), content.end(), '\n');
-  if (config.rank == 0) {
-    int listen_socket;
-    port_type port;
-    std::tie(listen_socket, port) = listen();
-
-    // pack message for other workers (we are first so we are master)
-    file << group_name << ' ' << port << ' ';
-    for (auto addr_str : getInterfaceAddresses()) {
-        file << addr_str << ' ';
-    }
-    file << std::endl;
+  // remember our order number
+  std::size_t order = std::count(content.begin(), content.end(), '\n');
+
+  int listen_socket;
+  port_type port;
+  std::tie(listen_socket, port) = listen();
+
+  auto ifa_addresses = getInterfaceAddresses();
+  file << group_name << ' ' << assigned_rank << ' ' << port
+       << ' ' << ifa_addresses.size();
+  for (auto addr_str : ifa_addresses) {
+    file << ' ' << addr_str;
+  }
+  file << std::endl;
 
-    file.close();
+  std::size_t lines = order + 1; // we have just added new line
+  while (lines < world_size) { // wait until all processes will write their info
     unlockFile(fd);
+    std::this_thread::sleep_for(std::chrono::milliseconds(200));
+    lockFile(fd);
+
+    file = std::fstream(file_path);
+    content = {std::istreambuf_iterator<char>(file),
+               std::istreambuf_iterator<char>()};
+    lines = std::count(content.begin(), content.end(), '\n');
+  }
+
+  file.seekp(0, std::ios_base::beg); // move read pos to beginning
+
+  std::vector<int> ranks(world_size);
+  std::vector<std::string> addresses;
+  for (std::size_t i = 0; i < world_size; ++i) {
+    std::string file_group_name, address;
+    std::size_t addresses_count;
+    int rank;
+    port_type tmp_port;
+
+    file >> file_group_name >> rank >> tmp_port >> addresses_count;
+    if (file_group_name != group_name) {
+      throw std::logic_error("file_group_name != group_name");
+    }
+
+    std::vector<std::string> tmp_addresses(addresses_count);
+    for (std::size_t j = 0; j < addresses_count; ++j) {
+      file >> address;
+      tmp_addresses.emplace(tmp_addresses.begin() + j, address);
+    }
+
+    ranks[i] = rank;
+    // Whether there is already assigned rank 0, or we have to get addresses and port
+    // from first process (which will be rank 0).
+    if (i == 0 || rank == 0) {
+      port = tmp_port;
+      addresses = tmp_addresses;
+    }
+  }
+
+  if (assigned_rank >= 0) {
+    config.rank = assigned_rank;
+  } else {
+    // Calculate how many unassigned there was before us
+    std::size_t unassigned = 1;
+    for (std::size_t i = 0; i < order; ++i) {
+      if (ranks[i] < 0)
+        unassigned++;
+    }
+
+    // Calculate actual rank by finding `unassigned` number of empty ranks
+    for (std::size_t rank = 0; rank < world_size; ++rank) {
+      if (std::find(ranks.begin(), ranks.end(), rank) == ranks.end()) {
+        unassigned--;
+      }
 
+      if (unassigned == 0) {
+        config.rank = rank;
+        break;
+      }
+    }
+  }
+
+  file << std::endl; // reserve our rank
```

Comment:
In no way does this reserve a rank. It only lets others know that we've finished the processing. It's best to connect this line with the `if` below, because it's only purpose is to tell when to remove the file